filebeat/input/filestream/input_test.go

```diff
@@ -0,0 +1,332 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filestream
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	loginp "github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile"
+	input "github.com/elastic/beats/v7/filebeat/input/v2"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/statestore"
+	"github.com/elastic/beats/v7/libbeat/statestore/backend/memlog"
+)
+
+// test_close_renamed from test_harvester.py
+func TestFilestreamCloseRenamed(t *testing.T) {
+	tmpDir, err := ioutil.TempDir("", "TestFilestreamCloseRenamed")
+	if err != nil {
+		t.Fatalf("cannot create temporary folder for test file: %+v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	testlogPath := filepath.Join(tmpDir, "test.log")
+	err = ioutil.WriteFile(testlogPath, []byte("first log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
+
+	testStore := openTestStatestore(t, tmpDir)
+	inp := getInputFromConfig(
+		t,
+		testStore,
+		map[string]interface{}{
+			"paths":                                []string{testlogPath},
+			"prospector.scanner.check_interval":    "1s",
+			"close.on_state_change.check_interval": "1s",
+			"close.on_state_change.renamed":        "true",
+		},
+	)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	inputCtx := input.Context{Logger: logp.L(), Cancelation: ctx}
+	pipeline := &mockPipelineConnector{}
+
+	var wg sync.WaitGroup
+	go func() {
+		wg.Add(1)
+		defer wg.Done()
+
+		inp.Run(inputCtx, pipeline)
+	}()
+
+	// first event has made it successfully
+	waitUntilEventCount(1, pipeline)
+	// check registry
+	checkOffsetInRegistry(t, testStore, testlogPath, len([]byte("first log line\n")))
+
+	testlogPathRenamed := filepath.Join(tmpDir, "test.log.rotated")
+	err = os.Rename(testlogPath, testlogPathRenamed)
+	if err != nil {
+		t.Fatalf("failed to rename file '%s': %+v", testlogPath, err)
+	}
+
+	// wait enough time so renaming of file can be detected
+	time.Sleep(2 * time.Second)
+
+	err = ioutil.WriteFile(testlogPath, []byte("new first log line\nnew second log line\n"), 0644)
+	if err != nil {
+		t.Fatalf("cannot write log lines to test file: %+v", err)
+	}
```

Comment:
Maybe this is too much for this test case?Given the test name we just want to validate that the original harvester is closed when the file is renamed. Yet the test simulates a file rotation scenario. Although I agree we need that kind of test, I wonder if we want to have a separate set of tests that deal with different configuration-combinations and file rotation.

Reply:
Indeed. But in the first round of moving tests to Golang I would prefer to keep this functionality. This helps me keep track of what tests I have ported from the `log` input. When I move on to supporting various log rotation methods, I am OK with renaming or moving the test around.

Reply:
Maybe you want to add an 'indicator' comment that you do not forget this particular line/test. e.g. `// TODO: ...` or `XXX: ...`